File: 1334. Find the City With the Smallest Number of Neighbors at a Threshold Distance/1334. Find the City With the Smallest Number of Neighbors at a Threshold Distance.cpp
class Solution {
public:
    int findTheCity(int n, vector<vector<int>>& edges, int distanceThreshold) {
        vector<vector<int>> distGraph(n, vector<int>(n, INT_MAX));
        
        for (int city = 0; city < n; ++city) {
            distGraph[city][city] = 0; 
        }

        for (auto& edg : edges) {
            distGraph[edg[0]][edg[1]] = edg[2];
            distGraph[edg[1]][edg[0]] = edg[2];
        }

        for (int intermediateCity = 0; intermediateCity < n; ++intermediateCity) {
            for (int firstCity = 0; firstCity < n; ++firstCity) {
                for (int secondCity = 0; secondCity < n; ++secondCity) {
                    if (distGraph[firstCity][intermediateCity] != INT_MAX &&
                        distGraph[intermediateCity][secondCity] != INT_MAX) {
                        distGraph[firstCity][secondCity] = min(
                            distGraph[firstCity][secondCity],
                            distGraph[firstCity][intermediateCity] + distGraph[intermediateCity][secondCity]
                        );
                    }
                }
            }
        }

        int minReach = INT_MAX, answerCity = -1;

        for (int srcCity = 0; srcCity < n; ++srcCity) {
            int srcCityReach = 0;
            for (int destCity = 0; destCity < n; ++destCity) {
                if (srcCity != destCity && distGraph[srcCity][destCity] <= distanceThreshold) {
                    srcCityReach++;
                }
            }
            if (srcCityReach <= minReach) {
                minReach = srcCityReach;
                answerCity = srcCity;
            }
        }

        return answerCity;
    }
};

File: 1334. Find the City With the Smallest Number of Neighbors at a Threshold Distance/1334. Find the City With the Smallest Number of Neighbors at a Threshold Distance.py

class Solution:
    def findTheCity(self, n: int, edges: List[List[int]], distanceThreshold: int) -> int:
    
        matrix = [[10001 for _ in range(n)] for _ in range(n)]
        for a, b, w in edges:
            matrix[a][b] = w
            matrix[b][a] = w
        for k in range(n):
            for i in range(n):
                for j in range(n):
                    matrix[i][j] = min(matrix[i][j],
                                    matrix[i][k] + matrix[k][j]
                                    )
        ans = 0
        minN = n
        for i in range(n):
            c = 0
            for j in range(n):
                if i==j: continue
                if(matrix[i][j] <= distanceThreshold):
                    c+=1
            if c<=minN:
                minN=c
                ans=i
        return ans